### DISCUSSION TITLE
Despite openssl@3 is not installed, `brew --prefix openssl` shows openssl@3’s path

### DISCUSSION THREAD

**Discussion by urimaro** [Homebrew/discussions#2187] [Everyday usage]
Hi,

I've installed `openssl@1.1` and I've never installed `openssl@3`.
But `brew --prefix openssl` shows openssl@3’s path.
I don't know why the openssl@3's path is shown.

`ls -la /usr/local/opt` shows below.

```
:
lrwxr-xr-x    1 goh   admin    24  9  6 11:21 openldap -> ../Cellar/openldap/2.5.7
lrwxr-xr-x    1 goh   admin    28  9  6 11:20 openssl -> ../Cellar/openssl@1.1/1.1.1l
lrwxr-xr-x    1 goh   admin    28  9  6 11:20 openssl@1.1 -> ../Cellar/openssl@1.1/1.1.1l
lrwxr-xr-x    1 goh   admin    23  9 12 11:20 pango -> ../Cellar/pango/1.48.10
:
```

The result of `which openssl` is below.

```
> which openssl
/usr/local/opt/openssl@1.1/bin/openssl
```

I just noticed two things.

- `brew search openssl` shows `openssle@1.1` and `openssl@3` but `openssl` is not shown
- `brew info openssl` shows the same information as `openssl@3`

I guess that `brew update` might affect this.
I’m not in trouble with this, but I thought it is not usual behavior.

Does anyone have any idea?

**Comment by carlocab** [comment]
`openssl` is an [alias to `openssl@3`](https://github.com/Homebrew/homebrew-core/blob/d43f843b6775f4332d328cc8361ec1402139077b/Aliases/openssl), so `brew` parses `openssl` as a reference to `openssl@3` when looking for exact matches. `brew search` also looks for inexact matches, so `brew search openssl` also returns `openssl@1.1`.

**Reply by urimaro** [reply]
Thanks for the reply.
That's helpful to understand the relationship between `openssl` and `openssl@3`

But I wonder why `brew --prefix openssl` returns `openssl@3`'s path even though `openssl@3` is not installed and the symbolic link `/usr/local/opt/openssl` points `openssl@1.1`.

**Reply by SMillerDev** [reply]
Because as far as brew knows, `openssl` is `openssl@3`. You just didn't apply that change yet by doing `brew upgrade`.

**Reply by carlocab** [reply]
> But I wonder why `brew --prefix openssl` returns `openssl@3`'s path even though `openssl@3` is not installed and the symbolic link `/usr/local/opt/openssl` points `openssl@1.1`.

Because `brew --prefix foo` returns the prefix for the formula matching `foo` exactly. The only formula that matches `openssl` exactly is `openssl@3`.

See also:
```
❯ brew help --prefix
Usage: brew --prefix [--unbrewed] [--installed] [formula ...]

Display Homebrew's install path. Default:

  - macOS Intel: /usr/local
  - macOS ARM: /opt/homebrew
  - Linux: /home/linuxbrew/.linuxbrew

If formula is provided, display the location where formula is or would be
installed.
```

**Reply by urimaro** [reply]
Thanks for the replies.
I'm glad to understand `openssl` points at `openssl@3`

---

@SMillerDev 

> You just didn't apply that change yet by doing brew upgrade.

Hmm...
Once again, I have never installed `openssl@3` and I have only installed `openssl@1.1`.

```
> brew list | grep openssl
openssl@1.1
```

In addition, I execute `brew update` and `brew upgrade` every day.
I am sorry that I have not seen how `brew upgrade` is related yet.
Would what you are trying to say be the same as what @carlocab said?

---

@carlocab 

> Because `brew --prefix foo` returns the prefix for the formula matching foo exactly.

I understand that `brew --prefix foo` uses the matching formula earlier than the symbolic link.
Am I understanding correctly?

> If formula is provided, display the location where formula is or would be installed.

I have missed the phrase especially 'or would be installed'.
Thanks.

**Reply by carlocab** [reply]
> > Because `brew --prefix foo` returns the prefix for the formula matching foo exactly.
> 
> I understand that `brew --prefix foo` uses the matching formula earlier than the symbolic link. Am I understanding correctly?

I'm not sure. This description seems to allow for an alias and a (core) formula of the same name, but `brew` doesn't allow that because I don't think it would be able to resolve which one you're referring to. Every alias must be distinct from each formula named in `$(brew --repository homebrew/core)/Formula`.

When you do `brew --prefix foo`, `brew` looks up a formula whose name or alias is `foo`. There is only one formula whose name or alias is `openssl`, and that is `openssl@3`.
